perception/lidar_transfusion/lib/transfusion_trt.cpp

```diff
@@ -0,0 +1,262 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "lidar_transfusion/transfusion_trt.hpp"
+
+#include "lidar_transfusion/preprocess/preprocess_kernel.hpp"
+#include "lidar_transfusion/transfusion_config.hpp"
+
+#include <tier4_autoware_utils/math/constants.hpp>
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace lidar_transfusion
+{
+
+TransfusionTRT::TransfusionTRT(
+  const NetworkParam & network_param, const DensificationParam & densification_param,
+  const TransfusionConfig & config)
+: config_(config)
+{
+  network_trt_ptr_ = std::make_unique<NetworkTRT>(config_);
+
+  network_trt_ptr_->init(
+    network_param.onnx_path(), network_param.engine_path(), network_param.trt_precision());
+  vg_ptr_ = std::make_unique<VoxelGenerator>(densification_param, config_, stream_);
+  stop_watch_ptr_ = std::make_unique<tier4_autoware_utils::StopWatch<std::chrono::milliseconds>>();
+  stop_watch_ptr_->tic("processing/inner");
+  initPtr();
+
+  CHECK_CUDA_ERROR(cudaStreamCreate(&stream_));
+
+#ifdef CPU_PROCESSING
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("lidar_transfusion"), "Using CPU processing");
+#else
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("lidar_transfusion"), "Using GPU processing");
+#endif
+}
+
+TransfusionTRT::~TransfusionTRT()
+{
+  if (stream_) {
+    cudaStreamSynchronize(stream_);
+    cudaStreamDestroy(stream_);
+  }
+}
+
+void TransfusionTRT::initPtr()
+{
+  // point cloud to voxels
+  voxel_features_size_ = config_.max_voxels_ * config_.max_num_points_per_pillar_ * 5;
+  voxel_num_size_ = config_.max_voxels_;
+  voxel_idxs_size_ = config_.max_voxels_ * 4;
+
+  // output of TRT -- input of post-process
+  cls_size_ = config_.num_proposals_ * config_.num_classes_;
+  box_size_ = config_.num_proposals_ * config_.num_box_values_;
+  dir_cls_size_ = config_.num_proposals_ * 2;
+  cls_output_d_ = cuda::make_unique<float[]>(cls_size_);
+  box_output_d_ = cuda::make_unique<float[]>(box_size_);
+  dir_cls_output_d_ = cuda::make_unique<float[]>(dir_cls_size_);
+
+  params_input_d_ = cuda::make_unique<unsigned int>();
+  voxel_features_d_ = cuda::make_unique<float[]>(voxel_features_size_);
+  voxel_num_d_ = cuda::make_unique<unsigned int[]>(voxel_num_size_);
+  voxel_idxs_d_ = cuda::make_unique<unsigned int[]>(voxel_idxs_size_);
+  points_d_ = cuda::make_unique<float[]>(config_.cloud_capacity_ * config_.num_point_feature_size_);
+  pre_ptr_.reset(new PreprocessCuda(config_, stream_));
```

Comment:
In the header file I commented that these may be unique ptrs
In addition, please avoid using new and instead use std::make_...

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/48200d207d459c01f9fb7dc38b33b355bc1437f0